### PR TITLE
Implement DeepSeek V2

### PR DESCRIPTION
**Pull request by EricLBuehler**
This PR implements the DeepSeek V2 architecture.

### CODE REVIEW

File: candle-transformers/src/models/deepseek2.rs [hunk 1]
```diff
@@ -0,0 +1,1114 @@
...
+                // (n, e)
+                // Invert the mask
+                let tmp_scores = masked_fill(&score_mask, &(1. - &score_mask.ne(0.)?)?, 0.)?;
+                let TopKOutput { values, indices } = tmp_scores.topk_unsorted(self.top_k)?;
+                (values, indices)
+            }
+        };
+
+        if self.top_k > 1 && self.cfg.norm_topk_prob {
+            let denmoninator = (topk_weight.sum_keepdim(D::Minus1)? + 1e-20)?;
```
Comment (LaurentMazare) [inline candle-transformers/src/models/deepseek2.rs:841]: typo: denominator

File: candle-examples/Cargo.toml [hunk 2]
```diff
@@ -122,3 +122,7 @@ required-features = ["onnx"]
 [[example]]
 name = "colpali"
 required-features = ["pdf2image"]
+
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-examples/Cargo.toml]: Fixed.

File: candle-examples/examples/deepseekv2/main.rs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    let tokenizer = Tokenizer::from_file(tokenizer_filename).map_err(E::msg)?;
+
+    let start = std::time::Instant::now();
+    let config: DeepSeekV2Config = {
+        let config_file = repo.get("config.json")?;
+        serde_json::from_slice(&std::fs::read(config_file)?)?
+    };
+    let device = candle_examples::device(args.cpu)?;
+    let (model, device) = {
+        let dtype = if device.is_cuda() {
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-examples/examples/deepseekv2/main.rs]: Fixed, I flipped the logic so we only use `f16` on the cpu.

File: candle-transformers/src/models/deepseek2.rs [hunk 4]
```diff
@@ -0,0 +1,1114 @@
...
+        let sorted_indices_all = self.arg_sort_last_dim(false)?;
+        let topk_indices_sorted = sorted_indices_all
+            .narrow(D::Minus1, 0, topk)?
+            .contiguous()?;
+        let topk_values_sorted = self.gather(&topk_indices_sorted, D::Minus1)?;
+
+        // Reorder the indices ascending
+        let reorder_indices = topk_indices_sorted.arg_sort_last_dim(true)?;
+        let topk_indices_unsorted = topk_indices_sorted
+            .to_dtype(DType::F32)?
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: Fixed, this was unnecessary.

File: candle-transformers/src/models/deepseek2.rs [hunk 5]
```diff
@@ -0,0 +1,1114 @@
...
+    Su,
+    #[serde(alias = "yarn")]
+    Yarn,
+    #[serde(alias = "dynamic")]
+    Dynamic,
+    #[serde(alias = "linear")]
+    Linear,
+}
+
+impl FromStr for ScaledRopeType {
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: Absolutely, removed.

File: candle-transformers/src/models/deepseek2.rs [hunk 6]
```diff
@@ -0,0 +1,1114 @@
...
+    type Err = candle::Error;
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "su" | "longrope" => Ok(Self::Su),
+            "yarn" => Ok(Self::Yarn),
+            "linear" => Ok(Self::Linear),
+            "dynamic" => Ok(Self::Dynamic),
+            _ => Err(candle::Error::Msg(
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: I've removed this `impl` entirely

File: candle-transformers/src/models/deepseek2.rs [hunk 7]
```diff
@@ -0,0 +1,1114 @@
...
+
+    fn forward(
+        &mut self,
+        xs: &Tensor,
+        attention_mask: Option<&Tensor>,
+        seqlen_offset: usize,
+    ) -> Result<Tensor> {
+        let (bs, seq_len, _) = xs.dims3()?;
+
+        let mut q = self.q.forward(xs)?;
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: I switched to using local scopes where possible; however, I think some are necessary (rope, kv cache, etc).

File: candle-transformers/src/models/deepseek2.rs [hunk 8]
```diff
@@ -0,0 +1,1114 @@
...
+        let mut v = kv_split[1].clone();
+
+        (q_pe, k_pe) = self.rotary_emb.forward(&q_pe, &k_pe, seqlen_offset)?;
+
+        let mut q = Tensor::zeros(
+            (bs, self.cfg.num_attention_heads, seq_len, self.q_head_dim),
+            q_pe.dtype(),
+            q_pe.device(),
+        )?;
+        q = q.slice_assign(
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: Ah, nice catch! I think it would be D::Minus1 though?

File: candle-transformers/src/models/deepseek2.rs (hunk 1 above)
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: Thanks!

File: candle-transformers/src/models/deepseek2.rs [hunk 9]
```diff
@@ -0,0 +1,1114 @@
...
+        let topk_values_unsorted = topk_values_sorted.gather(&reorder_indices, D::Minus1)?;
+        Ok(TopKOutput {
+            values: topk_values_unsorted,
+            indices: topk_indices_unsorted,
+        })
+    }
+}
+
+pub trait SplitOp {
+    fn split<D: Dim>(&self, splits: &[usize], dim: D) -> Result<Vec<Tensor>>;
```
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs:126]: Actually, my idea was to maybe eventually migrate all the `...Op` traits in this file to `candle_nn` as they are both used in DeepSeek 3 (that PR duplicates these ops) and might be very useful.

What do you think?

[the remark this answers is not in the document]
Comment (LaurentMazare) [reply candle-transformers/src/models/deepseek2.rs:126]: I would rather keep `candle_nn` on the simpler side, here it seems that the ops can be just specialized to what is exactly required by the model, I would even suggest defining helper functions rather than going through traits, and the deepseek-v3 implementation can use some functions from the deepseek-v2 module.

File: candle-transformers/src/models/deepseek2.rs [hunk 10]
```diff
@@ -0,0 +1,1053 @@
...
+            }
+            TopkMethod::GroupLimitedGreedy => {
+                // (n, n_group)
+                let group_scores = scores
+                    .reshape((bs * seq_len, self.cfg.n_group, ()))?
+                    .max(D::Minus1)?;
+                // (n, topk_group)
+                let group_idx = scores.topk_unsorted(self.cfg.topk_group)?.indices;
+                // (n, n_group)
+                let mut group_mask = group_scores.zeros_like()?;
```
Comment (LaurentMazare) [inline candle-transformers/src/models/deepseek2.rs:755]: Cannot you just avoid this `mut` by chaining calls or using a local scope, seems fairly easy to do. Please also review the other remaining muts.

File: candle-transformers/src/models/deepseek2.rs [hunk 11]
```diff
@@ -0,0 +1,1053 @@
...
+        };
+
+        let kv_split = kv.split(&[self.cfg.qk_nope_head_dim, self.cfg.v_head_dim], D::Minus1)?;
+        let k_nope = kv_split[0].clone();
+        let mut v = kv_split[1].clone();
+
+        (q_pe, k_pe) = self.rotary_emb.forward(&q_pe, &k_pe, seqlen_offset)?;
+
+        let q = Tensor::cat(&[q_nope, q_pe], D::Minus1)?;
+        let mut k = Tensor::cat(&[k_nope, k_pe], D::Minus1)?;
```
Comment (guoqingbao) [inline candle-transformers/src/models/deepseek2.rs:641]: Hi @EricLBuehler I got the following error for running DeepSeek-V2-Lite-Chat model:
```
called `Result::unwrap()` on an `Err` value: APIError { data: "shape mismatch in cat for dim 1, shape for arg 1: [1, 16, 5, 128] shape for arg 2: [1, 1, 5, 64]" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I found that you have repeated the second dim of Tensor `k_pe` to match Tensor `k_nope` and padded Tensor `v` with zeros in mistral.rs but not over here. Are there something specials for this case? I also tested the mistral.rs implementation, while, the lite model gives random outputs.

Comment (guoqingbao) [inline candle-transformers/src/models/deepseek2.rs:641]: It can simply fix the problem by adding a dim repeat

```
let k_pe = k_pe.repeat((1, q.dim(1)?, 1, 1))?;
```

Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs:641]: Fixed that, thanks for finding this!

File: candle-transformers/src/models/deepseek2.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs:126]: Sounds great, let's keep this as-is then.

File: candle-transformers/src/models/deepseek2.rs (hunk 10 above)
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: I've removed this `mut`. I removed all other `mut` occurrences in the model execution path (unless completely impossible), too.

File: candle-transformers/src/models/deepseek2.rs (hunk 11 above)
Comment (EricLBuehler) [reply candle-transformers/src/models/deepseek2.rs]: Hi @guoqingbao!

> I found that you have repeated the second dim of Tensor k_pe to match Tensor k_nope and padded Tensor v with zeros in mistral.rs but not over here. Are there something specials for this case?

That case is only for the v head dim to match q/k as PagedAttention requires that (be sure to unpad too). We don't have PagedAttention in Candle (yet?), so this is not included.

> I also tested the mistral.rs implementation, while, the lite model gives random outputs.

Just checked it there with DS V2 Lite on Metal in mistral.rs, and it works. Was that failing?

**Comment by EricLBuehler** [comment]
@LaurentMazare I updated the model with some fixes and removed all the `mut` occurrences, except for in the MoE block where it is necessary. If you could review again, that would be great!

I tested the model, and it is working:
```
cargo run --example deepseekv2 --release --features metal -- --prompt "Recursive fibonacci code in Rust:" --which lite --sample-len 150   
```

**Review by LaurentMazare — APPROVED** [review]

**Comment by EricLBuehler** [comment]
Thank you! I'll update the DeepSeek V3 PR #2745.

I also have some MoE-specific optimizations which I'll be pushing shortly!
